src/autocomplete/databases/mongo/mongo-extract-commands.ts

```diff
@@ -0,0 +1,267 @@
+import json5 from 'json5';
+import {
+    CollectionOperationContext,
+    FilterModifierContext,
+    FindMethodContext,
+    FindMethodModifierContext,
+    HintModifierContext,
+    InsertOneMethodContext,
+    LimitModifierContext,
+    MaxModifierContext,
+    MinModifierContext,
+    MongoParser,
+    ReturnKeyModifierContext,
+    ShowRecordIdModifierContext,
+    SkipModifierContext,
+    SortModifierContext,
+} from './generated/MongoParser';
+import {MongoParserVisitor} from './generated/MongoParserVisitor';
+import {ParserSyntaxError, SqlErrorListener, createParser} from '../../shared';
+import {MongoLexer} from './generated/MongoLexer';
+import {getParseTree} from './mongo-autocomplete';
+
+type FindModifier =
+    | {
+          method:
+              | 'skip'
+              | 'filter'
+              | 'min'
+              | 'max'
+              | 'returnKey'
+              | 'showRecordId'
+              | 'limit'
+              | 'hint';
+          parameters: unknown;
+      }
+    | {method: 'sort'; parameters?: unknown; options?: unknown};
+
+interface FindCommand {
+    method: 'find';
+    collectionName: string;
+    parameters?: unknown;
+    options?: unknown;
+    explain?: {
+        parameters?: unknown;
+    };
+    modifiers: FindModifier[];
+}
+
+type Command =
+    | FindCommand
+    | {
+          method: 'insertOne';
+          collectionName: string;
+          document: unknown;
+          options: unknown;
+      };
+
+type OnParseCommand = (command: Command) => void;
+type OnParseCommandError = (error: ExpectedError | UnexpectedError) => void;
+
+export interface ExpectedError {
+    type: 'expected';
+    message: string;
+}
+
+export interface UnexpectedError {
+    type: 'unexpected';
+    message: unknown;
+}
+
+type ExtractMongoCommandsFromQueryResult =
+    | {commands: Command[]}
+    | {parseSyntaxErrors: ParserSyntaxError[]}
+    | {parseCommandsError: ExpectedError | UnexpectedError};
+
+function newExpectedError(message: string): ExpectedError {
+    return {
+        type: 'expected',
+        message,
+    };
+}
+
+function newUnexpectedError(message: unknown): UnexpectedError {
+    return {
+        type: 'unexpected',
+        message,
+    };
+}
+
+class CommandsVisitor extends MongoParserVisitor<unknown> {
+    onParseCommand: OnParseCommand;
+    onParseCommandError: OnParseCommandError;
```

Comment:
why not use SymbolTable or something with similar architecture? It will be consistent with our previous code and much cleaner imo than passing callbacks

I think either way it's better to store entities inside visitor and then retrieve them outside

Reply:
I didn't know that we can use them like this, thank you!